src/Components/WebAssembly/WebAssembly.Authentication/src/RemoteAuthenticatorViewCore.cs

```diff
@@ -235,23 +237,33 @@ private async Task ProcessLogIn(string returnUrl)
 
     private async Task ProcessLogInCallback()
     {
-        var url = Navigation.Uri;
-        var result = await AuthenticationService.CompleteSignInAsync(new RemoteAuthenticationContext<TAuthenticationState> { Url = url });
+        var result = await AuthenticationService.CompleteSignInAsync(
+            new RemoteAuthenticationContext<TAuthenticationState> { Url = Navigation.Uri });
         switch (result.Status)
         {
             case RemoteAuthenticationStatus.Redirect:
                 // There should not be any redirects as the only time CompleteSignInAsync finishes
                 // is when we are doing a redirect sign in flow.
                 throw new InvalidOperationException("Should not redirect.");
             case RemoteAuthenticationStatus.Success:
-                await OnLogInSucceeded.InvokeAsync(result.State);
-                NavigateToReturnUrl(GetReturnUrl(result.State));
+                Log.LoginRedirectCompletedSuccessfully(Logger);
+                if (OnLogInSucceeded.HasDelegate)
+                {
+                    Log.InvokingLoginCompletedCallback(Logger);
+                    await OnLogInSucceeded.InvokeAsync(result.State);
+                }
+                var redirectUrl = GetReturnUrl(result.State);
+                Log.NavigatingToUrl(Logger, redirectUrl);
+                Navigation.NavigateTo(redirectUrl, AuthenticationNavigationOptions);
                 break;
             case RemoteAuthenticationStatus.OperationCompleted:
                 break;
             case RemoteAuthenticationStatus.Failure:
-                var uri = Navigation.ToAbsoluteUri($"{ApplicationPaths.LogInFailedPath}?message={Uri.EscapeDataString(result.ErrorMessage)}").ToString();
-                NavigateToReturnUrl(uri);
+                Log.LoginCallbackFailed(Logger, result.ErrorMessage);
+                Log.NavigatingToUrl(Logger, ApplicationPaths.LogInFailedPath);
+                Navigation.NavigateTo(
+                    ApplicationPaths.LogInFailedPath,
+                    AuthenticationNavigationOptions with { HistoryEntryState = result.ErrorMessage });
```

Comment:
Congratulations! First usage of `with` in this codebase (at least that I've noticed)

Reply:
It is actually pretty nice because it avoids a lot of repetition and allows you to define common things in one place.